tests/integration.rs

```diff
@@ -0,0 +1,123 @@
+#[macro_use]
+mod utils;
+
+use std::{iter::once, path::PathBuf};
+
+use fs_err as fs;
+use parse_display::Display;
+use proptest::sample::size_range;
+use rand::{rngs::SmallRng, RngCore, SeedableRng};
+use tempfile::tempdir;
+use test_strategy::{proptest, Arbitrary};
+
+use crate::utils::{assert_same_directory, write_random_content};
+
+// tar and zip extensions
+#[derive(Arbitrary, Debug, Display)]
+#[display(style = "lowercase")]
+enum DirectoryExtension {
+    Tar,
+    Tbz,
+    Tgz,
+    Tlz4,
+    Tlz,
+    Tlzma,
+    Txz,
+    Zip,
+}
+
+// extensions of single file compression formats
+#[derive(Arbitrary, Debug, Display)]
+#[display(style = "lowercase")]
+enum FileExtension {
+    Bz,
+    Bz2,
+    Gz,
+    // Lz4, // broken
+    Lz,
+    Lzma,
+    Xz,
+}
+
+#[derive(Arbitrary, Debug, Display)]
+#[display("{0}")]
+enum Extension {
+    Directory(DirectoryExtension),
+    File(FileExtension),
+}
+
+// converts a list of extension structs to string
+fn merge_extensions(ext: impl ToString, exts: Vec<FileExtension>) -> String {
+    once(ext.to_string()).chain(exts.into_iter().map(|x| x.to_string())).collect::<Vec<_>>().join(".")
+}
+
+// create random nested directories and files under the specified directory
+fn create_random_files(dir: impl Into<PathBuf>, depth: u8, rng: &mut SmallRng) {
+    if depth == 0 {
+        return;
+    }
+
+    let dir = &dir.into();
+
+    // create 0 to 7 random files
+    for _ in 0..rng.next_u32() % 8 {
+        write_random_content(&mut tempfile::Builder::new().tempfile_in(dir).unwrap().keep().unwrap().0, rng);
+    }
+
+    // create more random files in 0 to 3 new directories
+    for _ in 0..rng.next_u32() % 4 {
+        create_random_files(&tempfile::tempdir_in(dir).unwrap().into_path(), depth - 1, rng);
+    }
+}
+
+// compress and decompress a single empty file
+#[proptest(cases = 512)]
+fn single_empty_file(ext: Extension, #[any(size_range(0..8).lift())] exts: Vec<FileExtension>) {
```

Comment:
I took some time to understand how proptest works, and it seems like `ext: Extension` and `Vec<FileExtension>` can create generic examples for each enum, right? I really like this addittion, but I personally had some trouble in understanding the tests, since they show up as regular tests in the test suite:

```
     Running tests/integration.rs (target/debug/deps/integration-86976fb3c0298888)

running 5 tests
test utils::src_is_not_tests - should panic ... ok
test utils::src_is_src ... ok
test single_empty_file ... ok
test single_file ... ok
test multiple_files ... ok
```

Seeing at first glance, it is hard to understand that we are in fact checking a lot of stuff here, would you know of a way in which we could check which arguments were passed to this function, as to understand better how it is working?

If, for example, proptest could write what it used to a file, that would already improve a lot the understandability, something akin to this:

```
$ cargo test — proptest_write_stuff — tests.txt

$ cat tests.txt

# single_empty_file arguments:
ext = "zip", 3, exts = vec["zip", "gz" "bz"];
ext = "tar", 2, exts = vec["tar", "gz"];
...

# multiple_files arguments:
ext = "zip", exts = vec["zip", "gz" "bz"], depth = 3;
ext = "tar", exts = vec["tar", "gz"], depth = 2;
...

```

Of course this is only an example, anything you can think of would be helpful

Reply:
I don't know if there is an easy way to put all the arguments into a file, though the error messages generated by proptest are pretty readable imo

Reply:
All we would need here are the error messages anyways, so resolving.